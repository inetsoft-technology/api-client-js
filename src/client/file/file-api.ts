/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * FileApi provides operations used to import, export, and migrate content.
 */
export interface FileApi {
   /**
    * Converts a library file or directory.
    *
    * @param extract true if the file should be extracted or false if the directory packaged.
    */
   convertLibrary(extract: boolean): Promise<void>;

   /**
    * Converts and indexed storage file or directory.
    *
    * @param extract   true if the file should be extracted or false if the directory packaged.
    * @param overwrite true to overwrite existing files.
    */
   convertStorage(extract: boolean, overwrite: boolean): Promise<void>;

   /**
    * Imports assets into the repository.
    *
    * @param data      the binary content of asset ZIP file.
    * @param excluded  the identifiers of the assets to be excluded from the import.
    * @param overwrite true to overwrite existing assets.
    */
   importAssets(data: ArrayBuffer, excluded: string[], overwrite: boolean): Promise<void>;

   /**
    * Exports assets from the repository.
    *
    * @param name     the name of the exported ZIP file.
    * @param includes the list of patterns matching assets to include.
    * @param excludes the list of patterns matching assets to exclude.
    *
    * @return the binary content of asset ZIP file.
    */
   exportAssets(name: string, includes: string[], excludes: string[]): Promise<ArrayBuffer>;

   /**
    * Uploads the local file system data space to the database data space or downloads the database
    * data space to the local file system data space.
    *
    * @param download  true to download the database the file system; false to upload the file
    *                  system to the database.
    * @param overwrite true to overwrite existing files; false to retain existing files.
    */
   copyDataSpace(download: boolean, overwrite: boolean): Promise<void>;
}
