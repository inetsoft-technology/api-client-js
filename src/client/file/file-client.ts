/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { fromByteArray } from "base64-js";
import { AbstractClient } from "../abstract-client";
import { HttpAdapter } from "../http-adapter";
import { FileApi } from "./file-api";

/**
 * @hidden
 */
export class FileClient extends AbstractClient implements FileApi {

   constructor(adapter: HttpAdapter<any>, token: () => Promise<string>, baseUrl: string) {
      super(adapter, token, baseUrl);
   }

   public convertLibrary(extract: boolean): Promise<void> {
      const data = { extract };
      const url = "api/public/files/convert-library";
      return this.doPost(url, data);
   }

   public convertStorage(extract: boolean, overwrite: boolean): Promise<void> {
      const data = { extract, overwrite };
      const url = "api/public/files/convert-storage";
      return this.doPost(url, data);
   }

   public importAssets(data: ArrayBuffer, excluded: string[], overwrite: boolean): Promise<void> {
      const encoded = fromByteArray(new Uint8Array(data));
      const body = { excluded, overwrite, data: encoded };
      const url = "api/public/files/import-assets";
      return this.doPost(url, body);
   }

   public exportAssets(name: string, includes: string[], excludes: string[]): Promise<ArrayBuffer> {
      const body = { name, includes, excludes };
      const url = "api/public/files/export-assets";
      return this.doPostBinary(url, body);
   }

   public copyDataSpace(download: boolean, overwrite: boolean): Promise<void> {
      const data = { download, overwrite };
      const url = "api/public/files/copy-data-space";
      return this.doPost(url, data);
   }
}
