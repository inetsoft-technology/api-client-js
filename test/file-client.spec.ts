/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FileClient } from "../src/client/file/file-client";

describe("FileClient", () => {
   let adapter: any;
   let client: FileClient;

   beforeEach(() => {
      adapter = {
         get: jest.fn(),
         post: jest.fn(),
         put: jest.fn(),
         delete: jest.fn(),
         map: jest.fn(),
         mapText: jest.fn(),
         mapBinary: jest.fn()
      };
      client = new FileClient(adapter, () => Promise.resolve("secret"), "http://localhost:8080/sree/");
   });

   describe("when convertLibrary() is called with extract set to true", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.convertLibrary(true).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/convert-library");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when convertLibrary() is called with extract set to false", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.convertLibrary(false).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/convert-library");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when convertStorage(true, true) is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.convertStorage(true, true).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/convert-storage");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when convertStorage(true, false) is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.convertStorage(true, false).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/convert-storage");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when convertStorage(false, true) is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.convertStorage(false, true).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/convert-storage");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when convertStorage(false, false) is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.convertStorage(false, false).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/convert-storage");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when importAssets() is called with overwrite set to true", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);
         return client.importAssets(createBinaryData(), ["a1", "a2"], true).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/import-assets");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when importAssets() is called with overwrite set to false", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);
         return client.importAssets(createBinaryData(), ["a1", "a2"], false).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/import-assets");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when exportAssets() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({ data: createBinaryData() }));
         adapter.mapBinary.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.exportAssets("assets", ["i1", "i2"], ["e1", "e2"]).then(result => {
            validateBinaryData(result);
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/export-assets");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when copyDataSpace(true, true) is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.copyDataSpace(true, true).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/copy-data-space");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when copyDataSpace(true, false) is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.copyDataSpace(true, false).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/copy-data-space");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when copyDataSpace(false, true) is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.copyDataSpace(false, true).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/copy-data-space");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when copyDataSpace(false, false) is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.copyDataSpace(false, false).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/files/copy-data-space");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });
});

function createBinaryData(): ArrayBuffer {
   const buffer = new ArrayBuffer(16);
   const array = new Int32Array(buffer);
   array[0] = 1;
   array[1] = 2;
   array[2] = 3;
   array[3] = 4;
   return buffer;
}

function validateBinaryData(buffer: ArrayBuffer): void {
   expect(buffer).toBeTruthy();
   expect(buffer.byteLength).toBe(16);
   const actual = new Int32Array(buffer);
   expect(actual[0]).toBe(1);
   expect(actual[1]).toBe(2);
   expect(actual[2]).toBe(3);
   expect(actual[3]).toBe(4);
}
