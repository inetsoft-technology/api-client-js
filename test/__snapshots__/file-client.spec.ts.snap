// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileClient when convertLibrary() is called with extract set to false then the adapter is called with the correct arguments 1`] = `
Object {
  "extract": false,
}
`;

exports[`FileClient when convertLibrary() is called with extract set to true then the adapter is called with the correct arguments 1`] = `
Object {
  "extract": true,
}
`;

exports[`FileClient when convertStorage(false, false) is called then the adapter is called with the correct arguments 1`] = `
Object {
  "extract": false,
  "overwrite": false,
}
`;

exports[`FileClient when convertStorage(false, true) is called then the adapter is called with the correct arguments 1`] = `
Object {
  "extract": false,
  "overwrite": true,
}
`;

exports[`FileClient when convertStorage(true, false) is called then the adapter is called with the correct arguments 1`] = `
Object {
  "extract": true,
  "overwrite": false,
}
`;

exports[`FileClient when convertStorage(true, true) is called then the adapter is called with the correct arguments 1`] = `
Object {
  "extract": true,
  "overwrite": true,
}
`;

exports[`FileClient when copyDataSpace(false, false) is called then the adapter is called with the correct arguments 1`] = `
Object {
  "download": false,
  "overwrite": false,
}
`;

exports[`FileClient when copyDataSpace(false, true) is called then the adapter is called with the correct arguments 1`] = `
Object {
  "download": false,
  "overwrite": true,
}
`;

exports[`FileClient when copyDataSpace(true, false) is called then the adapter is called with the correct arguments 1`] = `
Object {
  "download": true,
  "overwrite": false,
}
`;

exports[`FileClient when copyDataSpace(true, true) is called then the adapter is called with the correct arguments 1`] = `
Object {
  "download": true,
  "overwrite": true,
}
`;

exports[`FileClient when exportAssets() is called then the adapter is called with the correct arguments 1`] = `
Object {
  "excludes": Array [
    "e1",
    "e2",
  ],
  "includes": Array [
    "i1",
    "i2",
  ],
  "name": "assets",
}
`;

exports[`FileClient when importAssets() is called with overwrite set to false then the adapter is called with the correct arguments 1`] = `
Object {
  "data": "AQAAAAIAAAADAAAABAAAAA==",
  "excluded": Array [
    "a1",
    "a2",
  ],
  "overwrite": false,
}
`;

exports[`FileClient when importAssets() is called with overwrite set to true then the adapter is called with the correct arguments 1`] = `
Object {
  "data": "AQAAAAIAAAADAAAABAAAAA==",
  "excluded": Array [
    "a1",
    "a2",
  ],
  "overwrite": true,
}
`;
